File: beakjoon-c,cpp/유기농 배추.cpp
#include<iostream>
#include<queue>
#include<vector>

using namespace std;

typedef pair<int,int> point;
vector<vector<bool>> check;
queue<point> q;
int dx[4] = {1,-1,0,0}, dy[4] = {0,0,1,-1};
int n,m,k,count = 0;

int BFS(int x,int y)
{
	check[x][y] = false;
	q.push({x,y});
	while(!q.empty())
	{
		x = q.front().first;
		y = q.front().second;
		q.pop();
		for(int i = 0; i < 4; i++)
		{
			int px = x + dx[i];
			int py = y + dy[i];
			if((0 > px || px >= n) || (0 > py || py >= m))
				continue;
				
			if(!check[px][py])
				continue;
			
			if(check[px][py])
			{
				check[px][py] = false;
				q.push({px,py});
			}
				
		}
	}
	return 1;
}
int main()
{
	int T;
	scanf("%d",&T);
	
	for(int i = 0; i < T; i++)
	{
		scanf("%d %d %d",&m,&n,&k);
		check.assign(n+2, vector<bool>(m+2,false));
		for(int j = 0; j < k; j++)
		{
			int x,y;
			scanf("%d %d",&y,&x);
			check[x][y] = true;
		}
		
		for(int p = 0; p < n; p++){
			for(int z = 0; z < m; z++)
			{
				if(check[p][z])
					count += BFS(p,z);
				else continue;
			}
		}
		printf("%d\n",count);
		count = 0;
		check.clear();
	}
		
	return 0;
} 

File: beckjoon-pyhton/유기농 배추.py
def BFS(check,x,y,n,m):
    check[x][y] = False
    q = [[x,y]]
    dx = [1,-1,0,0]
    dy = [0,0,1,-1]
    while(len(q)):
        x = q[0][0]
        y = q[0][1]
        q.pop(0)
        for i in range(4):
            px = x + dx[i]
            py = y + dy[i]
            if 0 > px or px >= n or 0 > py or py >= m:
                continue

            if not check[px][py]:
                continue

            if check[px][py]:
                check[px][py] = False
                q.append([px,py])
    return 1
    

def main():
    T = int(input())
    for i in range(T):
        m,n,k = map(int,input().split())
        check = [[False]*(m+1) for _ in range(n+1)]
        for j in range(k):
            x,y = map(int,input().split())
            check[y][x] = True
        
        result = 0
        for k in range(n):
            for j in range(m):
                if check[k][j]:
                    result += BFS(check,k,j,n,m)
        print(result)
        check.clear()
        result = 0
    return

main()